### PR TITLE
Adding OPA debugger

### PR DESCRIPTION
**Pull request by johanfylling**

### CODE REVIEW

File: src/da/activate.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  }
+}
+
+class OpaDebugAdapterExecutableFactory implements vscode.DebugAdapterDescriptorFactory {
+  createDebugAdapterDescriptor(
+    _session: vscode.DebugSession,
+    executable: vscode.DebugAdapterExecutable | undefined,
+  ): vscode.ProviderResult<vscode.DebugAdapterDescriptor> {
+    if (!executable) {
+      promptForUpdateRegal(minimumSupportedRegalVersion);
```
Comment (johanfylling) [inline src/da/activate.ts:121]: Do we need to allow the user to configure another debug adapter, similar to the `opa.languageServers` option for language servers?

File: package.json [hunk 2]
```diff
@@ -192,6 +216,88 @@
...
+        "languages": [
+          "rego"
+        ],
+        "configurationAttributes": {
+          "launch": {
+            "properties": {
+              "command": {
+                "type": "string",
+                "description": "The OPA command to run. E.g. 'eval', 'test'.",
+                "default": "run"
```
Comment (anderseknert) [inline package.json:238]: The default is `run`, as in `opa run`?

File: package.json [hunk 3]
```diff
@@ -192,6 +216,88 @@
...
+                "items": {
+                  "type": "string"
+                }
+              },
+              "bundlePaths": {
+                "type": "array",
+                "items": {
+                  "type": "string"
+                },
+                "default": ["${workspaceFolder}"]
```
Comment (anderseknert) [inline package.json:255]: Can this depend on another config attribute? If so, `opa.roots` would be a good choice in order to not have users repeating that here.

File: package.json [hunk 4]
```diff
@@ -192,6 +216,88 @@
...
+                "default": true
+              },
+              "enablePrint": {
+                "type": "boolean",
+                "description": "Enable print statements.",
+                "default": true
+              },
+              "logLevel": {
+                "type": "string",
+                "description": "Set the log level. One of: 'debug', 'info', 'warn', 'error'."
```
Comment (anderseknert) [inline package.json:290]: Log level of what exactly? :)
And what'd the default?

File: src/da/activate.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      }
+      if (targetResource) {
+        vscode.debug.startDebugging(undefined, {
+          type: "opa-debug",
+          name: "Debug Workspace",
+          request: "launch",
+          command: "eval",
+          query: "data",
+          bundlePaths: ["${workspaceFolder}"],
+          enablePrint: true,
```
Comment (anderseknert) [inline src/da/activate.ts:23]: Shouldn't these values be fetched from the config? Or that's not been resolved at this point?

Same question applies below where similar objects are returned.

File: src/da/activate.ts (hunk 1 above)
Comment (anderseknert) [reply src/da/activate.ts:121]: If someone comes along with a different implementation, we can accomodate that then. For now, I'd say we leave it be.

**Review by anderseknert — APPROVED** [review]
LGTM! Just a couple of questions.

File: package.json (hunk 3 above)
Comment (johanfylling) [reply package.json:255]: That should be possible to hook up, I think. We can't outright drop it and use `opa.roots` in its stead, as it must be possible to opt out this and only use the `dataPaths` setting. Would it make sense to allow you to specify `"${opa.roots}"`, and defaulting to `["$opa.roots"]`?

File: package.json (hunk 4 above)
Comment (johanfylling) [reply package.json:290]: The level of logging that is returned to the vs-code debug console. To make it possible to troubleshoot debugger issues. There is no default. If not set, no logging is returned at all. I'll update the description.

File: src/da/activate.ts (hunk 5 above)
Comment (johanfylling) [reply src/da/activate.ts:23]: The below objects are here to prime new launch configurations in `.vscode/launch.json` with default values.

This specific object is for when you launch an ad-hoc debug session with the debug button in the top right of the editor. I suppose we could allow you to configure this through `package.json`. But on the other hand, if you're not happy with the default values, you can just create a custom, names launch config in `.vscode/launch.json`.

File: package.json (hunk 4 above)
Comment (anderseknert) [reply package.json:290]: Thanks!

File: package.json (hunk 3 above)
Comment (anderseknert) [reply package.json:255]: yeah, as far as I understand the default value for `opa.roots` is already `["${workspaceFolder}"]`, so if you don't configure anything, it'd mean the same thing as above, while allowing you to set this via that option

File: src/da/activate.ts (hunk 5 above)
Comment (anderseknert) [reply src/da/activate.ts:23]: Just trying to ensure that we set as smart defaults as possible, and if you've already set e.g. bundlePaths in one location (your `.vscode/settings.json`) you shouldn't have to do it twice.

File: package.json (hunk 3 above)
Comment (johanfylling) [reply package.json:255]: Ah, so basically, if this isn't set, we use `opa.roots` by default. That makes sense 👍 .

File: src/da/activate.ts (hunk 5 above)
Comment (johanfylling) [reply src/da/activate.ts:23]: I think if I drop `bundlePaths` here, it should pan out to defaulting to whatever value you've set for `opa.roots` (once that change is in).

File: README.md [hunk 6]
```diff
@@ -32,6 +32,15 @@ To learn more about each language server feature, see the Regal [language server
...
 
+Regal also adds the ability to debug Rego modules via the [Debug Adapter Protocol](https://microsoft.github.io/debug-adapter-protocol/) (DAP):
+
+- Launching `eval` debug sessions
+- Setting and halting on breakpoints
+- Stepping over, into, and out of rules, functions, every-statements, and comprehensions
+- Inspecting the `data` and `input` documents
+- Inspecting local variable values
+- Inspecting values in the Virtual Cache (global results of rule and function evaluations)
+
```
Comment (johanfylling) [inline README.md:43]: We could also add a LSP-style gif animation here. But I have no idea how to do that 😄.

Comment (anderseknert) [reply README.md:43]: @charlieegan3 if you get some time later today or tomorrow, can you please try this out and make a recording dor the README in the process? :) We'll want to use that for the Regal docs too!

**Pull request merged**